File: test_dashboard_fix.js
#!/usr/bin/env node
/**
 * Test script to verify dashboard red flags fix
 */

console.log('🧪 Testing Dashboard Red Flags Fix');
console.log('=' * 40);

// Mock data to simulate API responses
const mockPolicies = [
  {
    id: 'policy-1',
    policy_name: 'GoldCare Employee Plan',
    policy_type: 'health',
    created_at: '2024-01-15T10:00:00Z'
  }
];

const mockRedFlags = [
  {
    id: 'flag-1',
    policy_id: 'policy-1',
    severity: 'high',
    title: 'Maternity Waiting Period',
    description: '12-month waiting period for maternity benefits'
  },
  {
    id: 'flag-2', 
    policy_id: 'policy-1',
    severity: 'high',
    title: 'Mental Health Visit Limitation',
    description: 'Limited to 12 mental health visits per year'
  },
  {
    id: 'flag-3',
    policy_id: 'policy-1', 
    severity: 'high',
    title: 'Out-of-Network Authorization',
    description: 'Prior authorization required for out-of-network services'
  },
  {
    id: 'flag-4',
    policy_id: 'policy-1',
    severity: 'medium',
    title: 'Pre-authorization Required',
    description: 'Pre-authorization required for specialist visits'
  },
  {
    id: 'flag-5',
    policy_id: 'policy-1',
    severity: 'medium',
    title: 'Coverage Limitation',
    description: 'Limited coverage for certain procedures'
  },
  {
    id: 'flag-6',
    policy_id: 'policy-1',
    severity: 'medium',
    title: 'Network Restriction',
    description: 'Restricted to specific provider network'
  },
  {
    id: 'flag-7',
    policy_id: 'policy-1',
    severity: 'medium',
    title: 'Exclusion',
    description: 'Cosmetic procedures excluded from coverage'
  },
  {
    id: 'flag-8',
    policy_id: 'policy-1',
    severity: 'medium',
    title: 'Waiting Period',
    description: 'Waiting period for certain benefits'
  },
  {
    id: 'flag-9',
    policy_id: 'policy-1',
    severity: 'low',
    title: 'Documentation Required',
    description: 'Additional documentation required for claims'
  }
];

// Simulate the dashboard API logic
function simulateDashboardStats() {
  console.log('\n📊 Simulating Dashboard API Logic...');
  
  // Calculate red flags summary (this is the fix we implemented)
  const redFlagsBySeverity = mockRedFlags.reduce((acc, flag) => {
    const severity = flag.severity || 'unknown';
    acc[severity] = (acc[severity] || 0) + 1;
    return acc;
  }, {});
  
  const dashboardStats = {
    total_policies: mockPolicies.length,
    total_documents: 1,
    policies_by_type: { health: 1 },
    policies_by_carrier: { 'GoldCare': 1 },
    recent_activity: [],
    red_flags_summary: {
      total: mockRedFlags.length,
      by_severity: redFlagsBySeverity
    }
  };
  
  return dashboardStats;
}

// Test the fix
const stats = simulateDashboardStats();

console.log('\n✅ Dashboard Stats Generated:');
console.log(`   Total Policies: ${stats.total_policies}`);
console.log(`   Total Documents: ${stats.total_documents}`);
console.log(`   Red Flags Total: ${stats.red_flags_summary.total}`);
console.log('\n🚩 Red Flags by Severity:');
Object.entries(stats.red_flags_summary.by_severity).forEach(([severity, count]) => {
  console.log(`   ${severity}: ${count}`);
});

// Verify the fix
const expectedTotal = 9;
const actualTotal = stats.red_flags_summary.total;

console.log('\n🧪 Test Results:');
if (actualTotal === expectedTotal) {
  console.log(`✅ PASS: Red flags count correct (${actualTotal}/${expectedTotal})`);
} else {
  console.log(`❌ FAIL: Red flags count incorrect (${actualTotal}/${expectedTotal})`);
}

// Check severity breakdown
const expectedSeverities = { high: 3, medium: 5, low: 1 };
let severityTestPassed = true;

Object.entries(expectedSeverities).forEach(([severity, expectedCount]) => {
  const actualCount = stats.red_flags_summary.by_severity[severity] || 0;
  if (actualCount !== expectedCount) {
    console.log(`❌ FAIL: ${severity} severity count incorrect (${actualCount}/${expectedCount})`);
    severityTestPassed = false;
  } else {
    console.log(`✅ PASS: ${severity} severity count correct (${actualCount})`);
  }
});

console.log('\n📋 Summary:');
console.log(`   Total Red Flags: ${severityTestPassed ? '✅ PASS' : '❌ FAIL'}`);
console.log(`   Severity Breakdown: ${severityTestPassed ? '✅ PASS' : '❌ FAIL'}`);

if (actualTotal === expectedTotal && severityTestPassed) {
  console.log('\n🎉 All tests passed! Dashboard fix is working correctly.');
  console.log('\n🚀 Next Steps:');
  console.log('   1. Start backend: python -m uvicorn app.main:app --reload');
  console.log('   2. Start frontend: npm run dev');
  console.log('   3. Login and check dashboard at http://localhost:3000/dashboard');
  console.log('   4. Verify red flags count shows 9 instead of 0');
} else {
  console.log('\n❌ Tests failed! Check the dashboard API implementation.');
}

File: test_dashboard_fix.py
#!/usr/bin/env python3
"""
Test the fixed dashboard API
"""

import sys
import os
sys.path.append(os.path.join(os.path.dirname(__file__), 'backend'))

def test_dashboard_api():
    print('🔧 Testing Fixed Dashboard API')
    print('=' * 40)

    try:
        from backend.app.services.policy_service import get_dashboard_summary_optimized
        from backend.app.utils.db import SessionLocal
        from backend.app.models.user import User
        from backend.app.schemas.dashboard import DashboardSummary, DashboardPolicy, DashboardRedFlag
        import json
        
        # Test the dashboard function
        db = SessionLocal()
        try:
            user = db.query(User).first()
            if user:
                print(f'Testing with user: {user.email}')
                
                # Test the optimized dashboard query
                dashboard_stats = get_dashboard_summary_optimized(db, user.id)
                print(f'✅ Dashboard stats query successful')
                print(f'   Total policies: {dashboard_stats["total_policies"]}')
                print(f'   Total documents: {dashboard_stats["total_documents"]}')
                print(f'   Total red flags: {dashboard_stats["red_flags_summary"]["total"]}')
                
                # Test schema validation
                test_policy = DashboardPolicy(
                    id="test-id",
                    policy_name="Test Policy",
                    policy_type="health",
                    created_at="2024-01-01T00:00:00",
                    carrier_name="Test Carrier",
                    carrier_code="TC"
                )
                print(f'✅ DashboardPolicy schema validation successful')
                
                test_red_flag = DashboardRedFlag(
                    id="test-id",
                    policy_id="policy-id",
                    title="Test Red Flag",
                    severity="high",
                    flag_type="test",
                    description="Test description",
                    created_at="2024-01-01T00:00:00",
                    policy_name="Test Policy"
                )
                print(f'✅ DashboardRedFlag schema validation successful')
                
                # Test JSON serialization
                policy_json = test_policy.model_dump()
                red_flag_json = test_red_flag.model_dump()
                json.dumps(policy_json)
                json.dumps(red_flag_json)
                print(f'✅ Schema objects are JSON serializable')
                
                return True
                
            else:
                print('❌ No users found')
                return False
                
        finally:
            db.close()
            
    except Exception as e:
        print(f'❌ Error: {e}')
        import traceback
        traceback.print_exc()
        return False

if __name__ == "__main__":
    success = test_dashboard_api()
    if success:
        print('\n🎉 Dashboard API fix validation successful!')
    else:
        print('\n❌ Dashboard API fix validation failed!')
